### PR TITLE
#14675: Investigate improved accuracy of column shared exponents for affected data types

### LINKED ISSUE #14675: Investigate improved accuracy of column shared exponents for affected data types
Some data types share exponents across rows. That may affect accuracy when performing operations across columns where there is a large difference between the largest and smallest values.

Need to add functionality and test improved precision.

### PR DESCRIPTION
**Pull request by mgajewskiTT**
### Ticket
#14675 

### Problem description
TT-Metal's bfloat8_b data type uses shared exponents across rows, which can significantly impact accuracy when performing operations across columns where there's a large difference between largest and smallest values. The task was to:

- Investigate how shared row exponents affect column-wise operations precision
- Design experiments to test bfloat8_b accuracy under non-uniform distribution patterns
- Measure precision differences using PCC, allclose, and ULP metrics
- Test both single tile (32×32) and multi-tile scenarios (16×16 tiles of 32×32)
- Develop functionality to systematically evaluate and improve shared exponent precision

### What's changed
This PR delivers a Shared Exponent Precision Testing Suite that systematically evaluates bfloat8_b accuracy issues:

- Targeted Experiment Design
   - Non-uniform distribution patterns: Created 8 spatial patterns specifically designed to stress-test shared row exponents (column gradients, magnitude variations, outlier placements)
   - Multi-scale testing: Single tile (32×32) and multi-tile (512×512 = 16×16 tiles) configurations to test shared exponent behavior at different scales
   - Column-focused operations: Emphasis on reduction operations (sum, mean, max) across columns where shared row exponents cause precision issues

- Comprehensive Precision Measurement
   - PCC analysis: Pearson correlation coefficient to measure linear relationship degradation
   - Allclose validation: Binary pass/fail testing at multiple tolerance levels (1e-2, 1e-3)
   - ULP quantification: Units in Last Place analysis with percentile distributions for floating-point precision assessment
   - Statistical comparison: Direct bfloat8_b vs bfloat16 reference comparison across all test scenarios

- Systematic Problem Identification
   - Pattern impact analysis: Quantifies which spatial patterns cause the worst precision degradation due to shared exponents
   - Worst case identification: Automatically identifies the most problematic combinations of data patterns and operations
   - Automated reporting: Generates detailed analysis reports highlighting specific scenarios where shared exponents fail

### Observations

Tested configuration: `wormhole n150`, operations: ttnn.sum, ttnn.max, ttnn.mean, ttnn.matmul, ttnn.matmul (tiny tiles), shapes: single tile (32,32), multi tile (16tiles x 16 tiles), rectangular ([(32, 128), (128, 32), (64, 256)])

### Key Findings

**Most Critical Issue**: Matrix multiplication with tiny tiles (matmul_tt) shows catastrophic failure when transpose=True is enabled, producing results that are anti-correlated with expected values (negative PCC).

**Operation Performance Hierarchy**:
1. **Standard matmul**: Generally reliable (PCC > 0.999)
2. **Sum/Mean operations**: Good performance (PCC > 0.998) with specific exceptions
3. **Max operations**: Some edge cases with zero PCC
4. **Matmul_tt (tiny tiles)**: Severe failures with transpose enabled

**Configuration Impact**:
- Tile width significantly affects precision (16 > 32)
- Multi-tile shapes accumulate errors exponentially
- Transpose operations fundamentally break tiny-tile matrix multiplication

**Distribution Sensitivity**:
- High variance distributions (`normal_high_var_100`) amplify errors
- Skewed distributions compound precision loss
- Outlier-containing distributions trigger extreme ULP errors

### Performance Hierarchy
Analysis across 880 tests per pattern reveals clear performance stratification:

| Rank | Pattern | Avg PCC | Characteristics |
|------|---------|---------|----------------|
| 1 (Best) | tile_boundaries | 0.827 | Most stable, lowest absolute errors (avg: 2,456) |
| 2 | diagonal_gradient | 0.827 | Moderate stability, higher error variance |
| 3 | row_outliers | 0.827 | High variance but manageable |
| 4 | checkerboard | 0.827 | Structured pattern with predictable behavior |
| 5 | block_pattern | 0.826 | Extreme absolute errors (up to 80.5M) despite reasonable PCC |
| 6 | row_uniform | 0.826 | Generally stable performance |
| 7 | fa_rand_pattern | 0.824 | Random patterns show high variance |
| 8 (Worst) | column_gradient | 0.823 | Can produce negative correlations, 257B average ULP error |

### Critical Pattern Observations

**Column Gradient** - The Worst Performer:
- Lowest average PCC (0.823) with high variability (σ=0.373)
- Can produce catastrophic negative correlations (min: -0.287)
- Average ULP errors of 257 billion indicate severe precision loss
- Particularly problematic with high-variance distributions

**Block Pattern** - Extreme Absolute Errors:
- Maintains reasonable average PCC (0.826) but hides severe localized errors
- Average absolute error: 2.4M, maximum: 80.5M
- Highest average ULP errors (397 billion)
- Suggests error concentration in specific regions

**Pattern Categories**:
- **Gradient patterns** (column, diagonal): Consistently poor precision
- **Structured patterns** (tile_boundaries, row_uniform): Better stability
- **Random/outlier patterns**: Moderate performance with high variance

### Catastrophic Failures by Metric

**Worst PCC (Correlation) Cases**:
| Configuration | PCC | Details |
|---------------|-----|---------|
| single_tile + column_gradient + matmul_tt | -0.287 | Complete anti-correlation |
| single_tile + fa_rand_pattern + matmul_tt | -0.269 | Systematic wrong direction |
| single_tile + column_gradient + matmul_tt | -0.251 | High variance distribution |

All worst PCC cases involve:
- matmul_tt (tiny tiles) with transpose=True
- Single tile configurations
- Gradient or random patterns

**Worst Absolute Error Cases**:
| Configuration | Error | Details |
|---------------|-------|---------|
| multi_tile + block_pattern + matmul_tt | 80.5M | skewed_high_var_100_negative |
| multi_tile + block_pattern + matmul_tt | 80.1M | skewed_high_var_100 |
| multi_tile + block_pattern + matmul_tt | 79.7M | skewed_high_var_100 |

Pattern: Multi-tile configurations with block patterns and high-variance distributions

**Worst ULP Error Cases**:
| Configuration | ULP Error | Details |
|---------------|-----------|---------|
| multi_tile + block_pattern + matmul_tt | 118.7T | normal_with_outliers |
| multi_tile + row_outliers + matmul_tt | 70.4T | normal_with_outliers |
| multi_tile + block_pattern + matmul_tt | 61.6T | normal_0_1 |

**Worst Relative Error Cases**:
| Configuration | Relative Error | Details |
|---------------|----------------|---------|
| multi_tile + checkerboard + sum | 20.5T | axis=1 reduction |
| multi_tile + row_uniform + matmul_tt | 1.9T | transpose operation |
| multi_tile + checkerboard + mean | 40B | axis=1 reduction |

### Critical Risk Combinations

1. **multi_tile + block_pattern + high_variance** → Extreme absolute errors
2. **reduction_ops + checkerboard + skewed_negative** → Astronomical relative errors
3. **column_gradient + high_variance distributions** → Negative correlations

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by mgajewskiTT** [comment]
I'm not sure if this is the right place for the script — let me know if you think it would fit better somewhere else.

**Comment by bbradelTT** [comment]
1.
> I'm not sure if this is the right place for the script — let me know if you think it would fit better somewhere else.

I'm worried that in other places, it might not be visible. I'm okay putting this under the matmul directory since it might be good to include a few unit tests for column sharded exponent accuracy. Alternatively, maybe somewhere in tests/sweep_framework (would need to be careful about potentially updating .github/workflows/ttnn-run-sweeps.yaml) 

@ayerofieiev-tt any suggestions?

2. How long do the tests take to run?

3. Does switching to column sharded instead of row sharded improve the accuracy for those cases where row sharded is bad (e.g. the column gradiant)?

4. Would you be able to add the md files you generated to the issue itself?

**Comment by mgajewskiTT** [comment]
> How long do the tests take to run?

In the current configuration: `wormhole n150`, tested operations: `ttnn.sum`, `ttnn.max`, `ttnn.mean`, `ttnn.matmul`, `ttnn.matmul (tiny tiles)`, tested shapes: `single tile (32,32)`, `multi tile (16 tiles x 16 tiles)`, `rectangular ([(32, 128), (128, 32), (64, 256)])`, all available distributions and patterns tests take ~7min 20s. Please keep in mind that in the current version, it’s not integrated into any CI and doesn’t even follow a pytest structure — it’s just a standalone script.

> Does switching to column sharded instead of row sharded improve the accuracy for those cases where row sharded is bad (e.g. the column gradiant)?

Currently we don’t shard tensors, they are prepared here: 
```python
def _run_precision_test(
    input_tensor: torch.Tensor, device, operation: str = "sum", axis: int = 0, matmul_config: dict = None
) -> dict:
    """Run a single precision test comparing bfloat16 vs bfloat8_b results"""

    # Store original for reference
    torch_optional_second_tensor = None
    ttnn_optional_second_tensor = None
    torch_tensor_bf16 = input_tensor.bfloat16()
    if operation in [OperationType.MATMUL_KEY, OperationType.MATMUL_TT_KEY]:
        # For matmul, create a second tensor with compatible shape
        second_shape = (input_tensor.shape[1], input_tensor.shape[0])
        torch_optional_second_tensor = torch.randn(second_shape).bfloat16()

    # Convert to ttnn tensors
    ttnn_tensor_bf8_b = ttnn.from_torch(torch_tensor_bf16, device=device, layout=ttnn.Layout.TILE, dtype=ttnn.bfloat8_b)

    if operation == OperationType.MATMUL_KEY:
        ttnn_optional_second_tensor = ttnn.from_torch(
            torch_optional_second_tensor,
            device=device,
            layout=ttnn.Layout.TILE,
            dtype=ttnn.bfloat8_b,
        )
    elif operation == OperationType.MATMUL_TT_KEY:
        ttnn_optional_second_tensor = ttnn.from_torch(
            torch_optional_second_tensor,
            device=device,
            layout=ttnn.Layout.TILE,
            dtype=ttnn.bfloat8_b,
            tile=ttnn.Tile(
                (32, matmul_config[MatmulTTConfig.TILE_W_KEY]),
                transpose_tile=matmul_config[MatmulTTConfig.TRANSPOSE_KEY],
            ),
        )
```
I can perform some additional test for matmul and matmul tinytiles with tensor sharding. 

> Would you be able to add the md files you generated to the issue itself?

Done!

File: tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/generators.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+    Args:
+        shape: tuple of (height, width) for the desired tensor shape
+
+    Returns:
+        Dictionary of pattern generation functions
+    """
+
+    # Pattern 1: Large magnitude differences across columns
+    def column_magnitude_gradient(shape):
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/generators.py:178]: This is an interesting patten.

Please add a
reverse_column_magnitude_gradient:
```
        for col in range(shape[1]):
            tensor[:, col] *= 10 ** -((col / shape[1]) * 6 - 3)  # 10^-3 to 10^3
```

a row_magnitude_gradient pattern
```
        for row in range(shape[0]):
            tensor[row, :] *= 10 ** ((row / shape[1]) * 6 - 3)  # 10^-3 to 10^3
```

and a reverse_row_magnitude_gradient pattern
```
        for row in range(shape[0]):
            tensor[row, :] *= 10 ** -((row / shape[1]) * 6 - 3)  # 10^-3 to 10^3
```

File: tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/constants.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    RESULTS_DIRECTORY = "bfloat8_experiment_results"
+    RAW_RESULTS_JSON_FILENAME = "raw_results.json"
+    RAW_RESULTS_MARKDOWN_FILENAME = "raw_results.md"
+    WORST_CASES_ANALYSIS_FILENAME = "worst_cases_analysis.md"
+    PATTERN_IMPACT_ANALYSIS_FILENAME = "pattern_impact_analysis.md"
+
+
+class ShapeType:
+    SINGLE_TILE_KEY = "single_tile"
+    MULTI_TILE_KEY = "multi_tile"
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/constants.py]: Since the multi tile test has a fixed size, it would be good to print that in raw_results.md. Not sure if the key should have the shape or the Shape type output formatted string should have the shape.

File: tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/generators.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+              - 'fa_rand_aggressive': Aggressive mixed Gaussian with 1% large outliers
+
+    Note:
+        All distributions are designed to stress-test shared exponent precision
+        and identify potential accuracy issues in TT-Metal operations.
+    """
+    distributions = {}
+
+    # Normal 0-1
+    distributions["normal_0_1"] = torch.randn(shape)
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/generators.py:140]: Please add another distribution:
constant: `torch.ones(shape)`

**Review by bbradelTT — COMMENTED** [review]
Thanks @mgajewskiTT for uploading the md files to the issue. Looks like there's something off with transposed tiles. All pccs are below 0.5:
```
awk 'BEGIN {ok=0; bad=0;} /matmul_tt \| .. \| True/ { if ($8 > 0.5) ok++; else bad++; } END {print ok, bad;}' raw_results.md
0 1280
```
I think that'll need to be investigated separately.

**Comment by bbradelTT** [comment]
Created https://github.com/tenstorrent/tt-metal/issues/30427 to look into what's going on with transposed tiles.

File: tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/generators.py (hunk 3 above)
Comment (mgajewskiTT) [reply tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/generators.py:140]: Added.

File: tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/generators.py (hunk 1 above)
Comment (mgajewskiTT) [reply tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/generators.py:178]: Added.

File: tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/constants.py (hunk 2 above)
Comment (mgajewskiTT) [reply tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/constants.py]: Added in `runner.py` since we define the shape there.

**Comment by ayerofieiev-tt** [comment]
@tt-ai please review this

File: tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/runner.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    elif operation == OperationType.MAX_KEY:
+        return ttnn.max(tensor, dim=axis)
+    elif operation in [OperationType.MATMUL_KEY, OperationType.MATMUL_TT_KEY]:
+        return ttnn.matmul(tensor, optional_second_tensor)
+    elif operation == OperationType.SOFTMAX_KEY:
+        return ttnn.softmax(tensor, dim=axis)
+    elif operation == OperationType.CUMSUM_KEY:
+        return ttnn.cumsum(tensor, dim=axis)
+    else:
+        raise ValueError(f"Unknown operation: {operation}")
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/runner.py:146]: @tt-ai suggest a map Key, torch, TT-NN
high probability copy-pate mistakes

Comment (mgajewskiTT) [reply tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/runner.py:146]: Changed.

File: tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/constants.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
+
+# SPDX-License-Identifier: Apache-2.0
+
+
+class ResultKeys:
+    PCC_KEY = "pcc"
```
Comment (rfurko-tt) [inline tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/constants.py]: could you please remove "_KEY" everywhere?

Comment (mgajewskiTT) [reply tests/ttnn/unit_tests/operations/shared_exponent_precision_testing_suite/constants.py]: Changed.

**Review by bbradelTT — APPROVED** [review]
